api/receiver/receive_results.go

```diff
@@ -98,38 +107,65 @@ func handleURLPayload(a AppEngineAPI, uploader string, urls []string, extraParam
 
 	var payloadType string
 	gcs := make([]string, 0, len(urls))
+	errors := make(chan error, len(urls))
+	var wg sync.WaitGroup
+	wg.Add(len(urls))
 
 	if len(urls) > 1 {
 		payloadType = "multiple"
 		for i, u := range urls {
-			f, err := a.fetchURL(u)
-			if err != nil {
-				return nil, err
-			}
-			defer f.Close()
-			fileName := fmt.Sprintf("%s/%s/%d.json", uploader, id, i)
-			// TODO: Detect whether the fetched blob is gzipped.
-			gcsPath, err := a.uploadToGCS(fileName, f, true)
-			if err != nil {
-				return nil, err
-			}
+			gcsPath := fmt.Sprintf("/%s/%s/%s/%d.json", BufferBucket, uploader, id, i)
 			gcs = append(gcs, gcsPath)
+			go saveFileToGCS(a, errors, &wg, u, gcsPath)
 		}
 	} else {
 		payloadType = "single"
-		f, err := a.fetchURL(urls[0])
-		if err != nil {
-			return nil, err
-		}
-		defer f.Close()
-		fileName := fmt.Sprintf("%s/%s.json", uploader, id)
-		// TODO: Detect whether the fetched blob is gzipped.
-		gcsPath, err := a.uploadToGCS(fileName, f, true)
-		if err != nil {
-			return nil, err
-		}
+		gcsPath := fmt.Sprintf("/%s/%s/%s.json", BufferBucket, uploader, id)
 		gcs = append(gcs, gcsPath)
+		saveFileToGCS(a, errors, &wg, urls[0], gcsPath)
+	}
+
+	wg.Wait()
+	close(errors)
+	var errStr string
+	for err := range errors {
+		errStr += err.Error() + "\n"
+	}
+	if errStr != "" {
+		return nil, fmt.Errorf("error(s) occured when retrieving results from %s:\n%s", uploader, errStr)
 	}
 
 	return a.scheduleResultsTask(uploader, gcs, payloadType, extraParams)
 }
+
+func saveFileToGCS(a AppEngineAPI, e chan error, wg *sync.WaitGroup, url, gcsPath string) {
+	defer wg.Done()
+
+	f, err := fetchFile(a, url)
+	if err != nil {
+		e <- err
+		return
+	}
+	defer f.Close()
+	// TODO: Detect whether the fetched blob is gzipped.
+	if err := a.uploadToGCS(gcsPath, f, true); err != nil {
+		e <- err
+	}
+}
+
+func fetchFile(a AppEngineAPI, url string) (io.ReadCloser, error) {
+	// It is safe to reuse the request as long as we are not modifying it.
+	log := shared.GetLogger(a.Context())
+	sleep := time.Millisecond * 500
+	for retry := 0; retry < NumRetries; retry++ {
```

Comment:
I like https://github.com/avast/retry-go for this sort of thing, but I don't feel strongly about it. If you think using it would make the code clearer then you can make the change.